File: shared/src/commonTest/kotlin/space/denhan/extract/ExtractionReviewTest.kt
package space.denhan.extract

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ExtractionReviewTest {

    private fun fields(
        serviceName: String? = "Netflix",
        serviceNameRaw: String? = "Netflix Premium",
        amountMinor: Long? = 231_000,
        amountRaw: String? = "231.000đ",
        currencyCode: String? = "VND",
        dueDateIso: String? = "2026-09-01",
        dueDateRaw: String? = "01/09/2026",
        dateFormat: DateFormat = DateFormat.DMY,
        confidence: Confidence = Confidence.HIGH,
    ) = ExtractedFields(
        sourceType = SourceType.BILLING_EMAIL,
        serviceName = serviceName,
        serviceNameRaw = serviceNameRaw,
        amountMinor = amountMinor,
        amountRaw = amountRaw,
        currencyCode = currencyCode,
        currencySymbolRaw = "đ",
        billingCycle = BillingCycle.MONTHLY,
        billingCycleRaw = "hàng tháng",
        dueDateIso = dueDateIso,
        dueDateRaw = dueDateRaw,
        dateFormat = dateFormat,
        confidence = confidence,
    )

    @Test
    fun `a clean extraction needs no attention`() {
        val result = ExtractionReview.review(fields())
        assertTrue(result.warnings.isEmpty(), "unexpected: ${result.warnings}")
        assertFalse(result.requiresAttention)
    }

    // The verbatim-quote check. A value with nothing behind it is the model
    // inventing something, and this catches it mechanically.
    @Test
    fun `a value with no supporting quote is flagged`() {
        val result = ExtractionReview.review(fields(amountRaw = null))
        assertTrue(result.warnings.any { it is Warning.UnsupportedValue })
        assertTrue(result.requiresAttention)
    }

    @Test
    fun `an unsupported name is flagged`() {
        val result = ExtractionReview.review(fields(serviceNameRaw = ""))
        assertTrue(result.warnings.any { it is Warning.UnsupportedValue })
    }

    @Test
    fun `an unsupported date is flagged`() {
        val result = ExtractionReview.review(fields(dueDateRaw = null))
        assertTrue(result.warnings.any { it is Warning.UnsupportedValue })
    }

    // 03/04 cannot be resolved from the text, so the app must ask rather than
    // guess. This is the failure that silently shifts a reminder by weeks.
    @Test
    fun `an ambiguous date requires a human decision`() {
        val result = ExtractionReview.review(
            fields(dueDateIso = null, dueDateRaw = "03/04/2026", dateFormat = DateFormat.AMBIGUOUS),
        )
        assertTrue(result.warnings.any { it is Warning.AmbiguousDate })
        assertTrue(result.requiresAttention)
    }

    @Test
    fun `both readings of an ambiguous date are offered`() {
        val readings = ExtractionReview.ambiguousReadings("03/04/2026")
        assertEquals("ngày 3 tháng 4" to "ngày 4 tháng 3", readings)
    }

    @Test
    fun `a date that cannot be ambiguous yields no alternative readings`() {
        assertNull(ExtractionReview.ambiguousReadings("25/12/2026"), "25 cannot be a month")
        assertNull(ExtractionReview.ambiguousReadings("no date here"))
        assertNull(ExtractionReview.ambiguousReadings(null))
    }

    @Test
    fun `separators other than slash are handled`() {
        assertEquals("ngày 3 tháng 4" to "ngày 4 tháng 3", ExtractionReview.ambiguousReadings("03-04-2026"))
        assertEquals("ngày 3 tháng 4" to "ngày 4 tháng 3", ExtractionReview.ambiguousReadings("03.04.2026"))
    }

    @Test
    fun `a missing date is flagged unless the text genuinely had none`() {
        assertTrue(
            ExtractionReview.review(fields(dueDateIso = null, dateFormat = DateFormat.DMY))
                .warnings.any { it is Warning.MissingDate },
        )
        assertFalse(
            ExtractionReview.review(
                fields(dueDateIso = null, dueDateRaw = null, amountMinor = null, amountRaw = null, dateFormat = DateFormat.ABSENT),
            ).warnings.any { it is Warning.MissingDate },
        )
    }

    // A bare number could be dong or dollars, and being wrong is a 26,000x error.
    @Test
    fun `an amount with no currency is flagged`() {
        val result = ExtractionReview.review(fields(currencyCode = null))
        assertTrue(result.warnings.any { it is Warning.UnknownCurrency })
        assertTrue(result.requiresAttention)
    }

    @Test
    fun `low confidence always requires review`() {
        val result = ExtractionReview.review(fields(confidence = Confidence.LOW))
        assertTrue(result.warnings.any { it is Warning.LowConfidence })
        assertTrue(result.requiresAttention)
    }

    @Test
    fun `medium confidence alone does not block`() {
        assertFalse(ExtractionReview.review(fields(confidence = Confidence.MEDIUM)).requiresAttention)
    }

    @Test
    fun `every warning carries a Vietnamese message`() {
        val result = ExtractionReview.review(
            fields(serviceNameRaw = null, amountRaw = null, currencyCode = null, confidence = Confidence.LOW),
        )
        assertTrue(result.warnings.isNotEmpty())
        result.warnings.forEach { assertTrue(it.messageVi.isNotBlank(), "$it has no message") }
    }
}

File: test/unit/extract/extraction_review_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:subdock/extract/extraction_review.dart';
import 'package:subdock/extract/extraction_schema.dart';

void main() {
  ExtractedFields fields({
    String? serviceName = 'Netflix',
    String? serviceNameRaw = 'Netflix Premium',
    int? amountMinor = 231000,
    String? amountRaw = '231.000đ',
    String? currencyCode = 'VND',
    String? dueDateIso = '2026-09-01',
    String? dueDateRaw = '01/09/2026',
    DateFormatKind dateFormat = DateFormatKind.dmy,
    Confidence confidence = Confidence.high,
  }) {
    return ExtractedFields(
      sourceType: SourceType.billingEmail,
      serviceName: serviceName,
      serviceNameRaw: serviceNameRaw,
      amountMinor: amountMinor,
      amountRaw: amountRaw,
      currencyCode: currencyCode,
      currencySymbolRaw: 'đ',
      billingCycle: BillingCycle.monthly,
      billingCycleRaw: 'hàng tháng',
      dueDateIso: dueDateIso,
      dueDateRaw: dueDateRaw,
      dateFormat: dateFormat,
      confidence: confidence,
    );
  }

  test('a clean extraction needs no attention', () {
    final result = ExtractionReview.review(fields());
    expect(result.warnings, isEmpty, reason: 'unexpected: ${result.warnings}');
    expect(result.requiresAttention, isFalse);
  });

  // A value the model gave without quoting anything to support it is the model
  // inventing something. Cheap, mechanical falsifier.
  test('a value with no supporting quote is flagged', () {
    final result = ExtractionReview.review(fields(amountRaw: null));
    expect(result.warnings.whereType<UnsupportedValue>(), isNotEmpty);
    expect(result.requiresAttention, isTrue);
  });

  test('an unsupported name is flagged', () {
    final result = ExtractionReview.review(fields(serviceNameRaw: '  '));
    expect(
      result.warnings.whereType<UnsupportedValue>().map((w) => w.field),
      contains('service name'),
    );
  });

  test('an unsupported date is flagged', () {
    final result = ExtractionReview.review(fields(dueDateRaw: null));
    expect(
      result.warnings.whereType<UnsupportedValue>().map((w) => w.field),
      contains('due date'),
    );
  });

  // 03/04 cannot be resolved from the text at all, so the app must not pick.
  test('an ambiguous date requires a human decision', () {
    final result = ExtractionReview.review(
      fields(
        dueDateIso: null,
        dueDateRaw: '03/04/2026',
        dateFormat: DateFormatKind.ambiguous,
      ),
    );
    expect(result.warnings.whereType<AmbiguousDate>(), isNotEmpty);
    expect(result.requiresAttention, isTrue);
  });

  test('both readings of an ambiguous date are offered', () {
    final readings = ExtractionReview.ambiguousReadings('03/04/2026');
    expect(readings, isNotNull);
    expect(readings!.$1, '3 April');
    expect(readings.$2, '4 March');
  });

  test('a date that cannot be ambiguous yields no alternative readings', () {
    expect(ExtractionReview.ambiguousReadings('18/08/2026'), isNull);
    expect(ExtractionReview.ambiguousReadings(null), isNull);
    expect(ExtractionReview.ambiguousReadings('no digits here'), isNull);
  });

  test('separators other than slash are handled', () {
    expect(ExtractionReview.ambiguousReadings('03-04-2026'), isNotNull);
    expect(ExtractionReview.ambiguousReadings('03.04.2026'), isNotNull);
  });

  test('a missing date is flagged unless the text genuinely had none', () {
    final missing = ExtractionReview.review(
      fields(
        dueDateIso: null,
        dueDateRaw: null,
        dateFormat: DateFormatKind.dmy,
      ),
    );
    expect(missing.warnings.whereType<MissingDate>(), isNotEmpty);

    final none = ExtractionReview.review(
      fields(
        dueDateIso: null,
        dueDateRaw: null,
        dateFormat: DateFormatKind.absent,
      ),
    );
    expect(none.warnings.whereType<MissingDate>(), isEmpty);
  });

  test('an amount with no currency is flagged', () {
    final result = ExtractionReview.review(fields(currencyCode: null));
    expect(result.warnings.whereType<UnknownCurrency>(), isNotEmpty);
    expect(result.requiresAttention, isTrue);
  });

  test('low confidence always requires review', () {
    final result = ExtractionReview.review(fields(confidence: Confidence.low));
    expect(result.warnings.whereType<LowConfidence>(), isNotEmpty);
    expect(result.requiresAttention, isTrue);
  });

  test('medium confidence alone does not block', () {
    final result = ExtractionReview.review(
      fields(confidence: Confidence.medium),
    );
    expect(result.requiresAttention, isFalse);
  });

  // These strings are what the user reads on the review screen; an English
  // fallback slipping through would be a visible defect.
  test('every warning carries a Vietnamese message', () {
    const warnings = <Warning>[
      UnsupportedValue('x'),
      AmbiguousDate('03/04'),
      MissingDate(),
      UnknownCurrency('đ'),
      LowConfidence(),
    ];
    for (final warning in warnings) {
      expect(warning.message, isNotEmpty);
      expect(warning.blocksAutoConfirm, isTrue);
    }
  });

  group('parsing the model reply', () {
    test('every field being absent is a valid reply, not a parse error', () {
      final parsed = ExtractedFields.fromJson(const {
        'source_type': 'unknown',
        'service_name': null,
        'service_name_raw': null,
        'amount_minor': null,
        'amount_raw': null,
        'currency_code': null,
        'currency_symbol_raw': null,
        'billing_cycle': null,
        'billing_cycle_raw': null,
        'due_date_iso': null,
        'due_date_raw': null,
        'date_format_detected': 'absent',
        'confidence': 'low',
      });

      expect(parsed.dueDateIso, isNull);
      expect(parsed.amountMinor, isNull);
      expect(parsed.dateFormat, DateFormatKind.absent);
    });

    test('an unrecognised enum value is refused rather than guessed', () {
      expect(
        () => ExtractedFields.fromJson(const {
          'source_type': 'from_the_future',
          'date_format_detected': 'absent',
          'confidence': 'low',
        }),
        throwsFormatException,
      );
    });
  });
}
